### PR TITLE
Extend `device_scalar` to optionally use pinned bounce buffer

### PR DESCRIPTION
**Pull request by vuule**
## Description
Depends on https://github.com/rapidsai/cudf/pull/16945

Added `cudf::detail::device_scalar`, derived from `rmm::device_scalar`. The new class overrides function members that perform copies between host and device. New implementation uses a `cudf::detail::host_vector` as a bounce buffer to avoid performing a pageable copy.

Replaced `rmm::device_scalar` with `cudf::detail::device_scalar` across libcudf.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/include/cudf/detail/device_scalar.hpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+
+  void set_value_async(T const& value, rmm::cuda_stream_view stream)
+  {
+    bounce_buffer[0] = value;
+    cuda_memcpy_async<T>(device_span<T>{this->data(), 1}, bounce_buffer, stream);
+  }
+
+  void set_value_async(T&& value, rmm::cuda_stream_view stream)
+  {
+    bounce_buffer[0] = std::move(value);
```
Comment (vuule) [inline cpp/include/cudf/detail/device_scalar.hpp:81]: bonus feature from having a bounce buffer - we don't need to worry about the `value` lifetime.

**Comment by davidwendt** [comment]
Why the global change? Could we still use `rmm::device_scalar` when not using a pinned bounce buffer?

File: cpp/src/join/distinct_hash_join.cu [hunk 2]
```diff
@@ -17,6 +17,7 @@
 #include "join_common_utils.hpp"
 
 #include <cudf/detail/cuco_helpers.hpp>
+#include <cudf/detail/device_scalar.hpp>
```
Comment (lamarrr) [inline cpp/src/join/distinct_hash_join.cu:20]: seems this include is now redundant

File: cpp/include/cudf/detail/device_scalar.hpp [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+ public:
+#ifdef __CUDACC__
+#pragma nv_exec_check_disable
+#endif
+  ~device_scalar() = default;
+
+#ifdef __CUDACC__
+#pragma nv_exec_check_disable
+#endif
+  device_scalar(device_scalar&& other) noexcept
```
Comment (lamarrr) [inline cpp/include/cudf/detail/device_scalar.hpp:41]: we can use = default for this

**Review by lamarrr — APPROVED** [review]

File: cpp/include/cudf/detail/device_scalar.hpp (hunk 3 above)
Comment (vuule) [reply cpp/include/cudf/detail/device_scalar.hpp:41]: That's what I did initially but compiler ended up seemingly not generating the move ctor on CUDA 11.8 (error was that copy ctor is deleted). Worked fine on 12.5.
I should add a comment.

**Review by vyasr — COMMENTED** [review]
The new class seems fine, but do we really need to use it everywhere? Does this mean that any usage of rmm::device_scalar in cudf is now forbidden because it could introduce unexpected performance overheads? If so, should we include a pre-commit hook or something to that effect to enforce that? Also, if this implication is correct @davidwendt may want to weigh in. 

The title says "optionally use pinned bounce buffer", but this usage looks unconditional. Is the optionality encoded in the host vector, or is it simply no longer optional?

**Comment by vuule** [comment]
> The title says "optionally use pinned bounce buffer", but this usage looks unconditional. Is the optionality encoded in the host vector, or is it simply no longer optional?

The use of a bounce buffer is unconditional, but `host_vector`, which is used as a bounce buffer, optionally uses pinned memory. 
There's no perf impact when the bounce buffer is pageable. I'll evaluate impact from pinned memory once we're closer to eliminating all "unconditional" pageable memory use (for memory that ends up on the GPU, or copied to from the GPU).

**Comment by vuule** [comment]
> The new class seems fine, but do we really need to use it everywhere? Does this mean that any usage of rmm::device_scalar in cudf is now forbidden because it could introduce unexpected performance overheads?

Pretty much. My aim is to stop using rmm::device_scalar outside of public APIs. Same goes for `std::vector` and `thrust::host_vector` (if copied to/from GPU). I don't know if we can automate this.

**Review by vyasr — APPROVED** [review]
Approving based on some extensive offline discussion. We're going to move frward with this and see how this type works in libcudf, then consider upstreaming it to rmm if it's generalizable, and if it's not then looking into developing clear guidelines for when it should be used in cudf.

File: cpp/include/cudf/detail/device_scalar.hpp [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+// Implementation is the same as what compiler should generate
+// Could not use default move constructor as 11.8 compiler fails to generate it
+#ifdef __CUDACC__
+#pragma nv_exec_check_disable
+#endif
+  device_scalar(device_scalar&& other) noexcept
+    : rmm::device_scalar<T>{std::move(other)}, bounce_buffer{std::move(other.bounce_buffer)}
+  {
+  }
+  device_scalar& operator=(device_scalar&&) noexcept = default;
```
Comment (davidwendt) [inline cpp/include/cudf/detail/device_scalar.hpp:47]: Curious why the move ctor required code but this did not?

Comment (vuule) [reply cpp/include/cudf/detail/device_scalar.hpp:47]: Default implementations should be fine in both cases. Compiled fine on 12.5 🤷
I suspect it's an 11.8 compiler bug, but really didn't want to dig into it, with a handy workaround available.

**Review by davidwendt — APPROVED** [review]

**Pull request merged**